File: src/test/scala/gameover/fwk/libgdx/gfx/GraphicsLoaderTest.scala
package gameover.fwk.libgdx.gfx

import com.badlogic.gdx.graphics.TextureData
import com.badlogic.gdx.math.Rectangle
import gameover.fwk.gdx.GdxTest
import org.scalatest.FlatSpec

class GraphicsLoaderTest extends FlatSpec {

  GdxTest.initializeLibGdx()
  val graphicsLoader = new GraphicsLoader()

  "GraphicsLoader" should "scan files and load images" in {
    assert(graphicsLoader.size > 0)
  }

  it should "load properly an image as a pixmap and a texture" in {
    graphicsLoader.pixmap("pad_area") match {
      case Some(pixmap) =>
      case _ => fail("Pixmap is not found")
    }

    graphicsLoader.texture("pad_area") match {
      case Some(texture) =>
      case _ => fail("Texture is not found")
    }
  }

  it should "load a nine patch getting information from filename" in {
    graphicsLoader.ninePatch("bar_window") match {
      case Some(ninePatch) =>
      case _ => fail("Nine patch is not found")
    }

    graphicsLoader.texture("pad_area") match {
      case Some(texture) =>
      case _ => fail("Texture is not found")
    }
  }


  it should "load properly an animation with an ara, the accurate number of frames and duration" in {
    graphicsLoader.animation("animwitharea_move_right") match {
      case Some(anim) =>
        assert(anim.anim.getFrameDuration == 0.08f)
        assert(anim.anim.getKeyFrames.length == 4)
        assert(anim.optionalArea.isDefined)
        assert(anim.optionalArea.get == new Rectangle(5f, 5f, 3f, 3f))
      case _ => fail("Animation is not found")
    }
  }

  it should "load properly an animation without an area, the accurate number of frames and duration" in {
    graphicsLoader.animation("animwithoutarea_move_right") match {
      case Some(anim) =>
        assert(anim.anim.getFrameDuration == 0.08f)
        assert(anim.anim.getKeyFrames.length == 4)
        assert(anim.optionalArea.isEmpty)
      case _ => fail("Animation is not found")
    }
  }

  it should "create a 'stand' animation from a 'move' animation containing the first frame only" in {
    val moveAnimation = graphicsLoader.animation("animwitharea_move_right")
    assume(moveAnimation.isDefined)
    graphicsLoader.animation("animwitharea_stand_right") match {
      case Some(standAnim) =>
        assert(standAnim.anim.getKeyFrames.length == 1)
        val td1: TextureData = standAnim.anim.getKeyFrames()(0).getTexture.getTextureData
        td1.prepare()
        val td2: TextureData = moveAnimation.get.anim.getKeyFrames()(0).getTexture.getTextureData
        td2.prepare()
        assert(td1.consumePixmap().getPixels == td2.consumePixmap().getPixels)
        td1.disposePixmap()
        td2.disposePixmap()
      case _ => fail("Stand animation is not found")
    }
  }
}

File: src/test/kotlin/gameover/fwk/libgdx/gfx/GraphicsLoaderTest.kt
package gameover.fwk.libgdx.gfx

import com.badlogic.gdx.math.Rectangle
import gameover.fwk.libgdx.GdxTest
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestInstance

@TestInstance(TestInstance.Lifecycle.PER_CLASS)
internal class GraphicsLoaderTest {
    init {
        GdxTest.initializeLibGdx()
    }
    private var graphicsLoader = GraphicsLoader()

    @Test
    fun `scanning files`(){
        assertThat(graphicsLoader.size()).isGreaterThan(0)
    }

    @Test
    fun `load properly an image as a pixmap and as a texture`() {
        assertThat(graphicsLoader.pixmap("pad_area")).isNotNull
        assertThat(graphicsLoader.texture("pad_area")).isNotNull
    }

    @Test
    fun `load a nine patch with information from filename`() {
        val ninePatch = graphicsLoader.ninePatch("bar_window")
        assertThat(ninePatch).isNotNull
    }

    @Test
    fun `load animation with an area the nb of frames and the duration`() {
        val ai = graphicsLoader.animation("animwitharea_move_right")
        assertThat(ai).isNotNull
        assertThat(ai?.anim?.frameDuration).isEqualTo(0.08f)
        assertThat(ai?.anim?.keyFrames).hasSize(4)
        assertThat(ai?.optionalArea).isNotNull
        assertThat(ai?.optionalArea).isEqualTo(Rectangle(5f, 5f, 3f, 3f))

    }

    @Test
    fun `load anim without area but nb of frames and the duration`() {
        val ai = graphicsLoader.animation("name")
        assertThat(ai).isNotNull
        assertThat(ai?.anim?.frameDuration).isEqualTo(0.12f)
        assertThat(ai?.anim?.keyFrames).hasSize(7)
        assertThat(ai?.optionalArea).isNull()
    }

    @Test
    fun `create a 'stand' animation from a 'move' animation containing the first frame only`() {
        val move = graphicsLoader.animation("animwitharea_move_right")
        assertThat(move).isNotNull
        val stand = graphicsLoader.animation("animwitharea_stand_right")
        assertThat(stand?.anim?.keyFrames).hasSize(1)
        val td1 = stand?.anim!!.keyFrames[0].texture.textureData
        td1.prepare()
        val td2 = move?.anim!!.keyFrames[0].texture.textureData
        td2.prepare()
        try {
            assertThat(td1.consumePixmap().pixels).isEqualTo(td2.consumePixmap().pixels)
        } finally {
            td1.disposePixmap()
            td2.disposePixmap()
        }
    }
}